### DISCUSSION TITLE
Could not embed because more than one relationship was found for 'creator_ranks' and 'public_profile'

### DISCUSSION THREAD

**Discussion by Electric-Dragon** [supabase/supabase#6848] [Questions] [answered]
I am trying to query a foreign table through a view and the following error appears. `Could not embed because more than one relationship was found for 'creator_ranks' and 'public_profile'`

---

This is the view definition for `creator_ranks`

```
create or replace view creator_ranks as (
	SELECT
		public_profile.id as creator_id,
		RANK() OVER (
			ORDER BY ((0.3 * total_likes) + (0.5 * get_creator_follows(creator_id)) + (0.2 * get_creator_total_reads(creator_id)))
		) creator_rank

	FROM public_profile

	JOIN creator_total_likes on public_profile.id = creator_total_likes.creator_id

)
```

Definition for `public_profile`:
```
CREATE TABLE IF NOT EXISTS public.public_profile
(
    id uuid NOT NULL,
    pfp text COLLATE pg_catalog."default" NOT NULL DEFAULT 'https://zduoakccvlarenfacitl.supabase.in/storage/v1/object/public/users/defaultpfp.jpeg'::text,
    username character varying COLLATE pg_catalog."default",
    description character varying COLLATE pg_catalog."default" DEFAULT ' '::character varying,
    instagram character varying COLLATE pg_catalog."default",
    reddit character varying COLLATE pg_catalog."default",
    youtube character varying COLLATE pg_catalog."default",
    banner text COLLATE pg_catalog."default",
    creator_type text[] COLLATE pg_catalog."default",
    is_creator boolean NOT NULL DEFAULT false,
    sample_arts text[] COLLATE pg_catalog."default",
    verified boolean NOT NULL DEFAULT false,
    CONSTRAINT public_profile_pkey PRIMARY KEY (id),
    CONSTRAINT public_profile_username_key UNIQUE (username),
    CONSTRAINT public_profile_id_fkey FOREIGN KEY (id)
        REFERENCES auth.users (id) MATCH SIMPLE
        ON UPDATE NO ACTION
        ON DELETE NO ACTION
)
```

---

This is the query I am making to supabase:
```
         const { data:creators, error:creatorsError } = await supabase
             .from('creator_ranks')
             .select('public_profile!inner(username,pfp,creator_type)')
             .order('creator_rank', { ascending: false })
             .limit(5)
```

---

This is `creatorsError.details`:

```
[Log] Array (4) (home.js, line 139)
0 {relationship: "public.adaptation_notifications[adaptation_notific…rom_id_fkey][adaptation_notifications_to_id_fkey]", embedding: "creator_ranks with public_profile", cardinality: "many-to-many"}
1 {relationship: "public.adaptation_notifications[adaptation_notific…o_id_fkey][adaptation_notifications_from_id_fkey]", embedding: "creator_ranks with public_profile", cardinality: "many-to-many"}
2 {relationship: "public.creator_follows[creator_follows_follower_fkey][creator_follows_following_fkey]", embedding: "creator_ranks with public_profile", cardinality: "many-to-many"}
3 {relationship: "public.creator_follows[creator_follows_following_fkey][creator_follows_follower_fkey]", embedding: "creator_ranks with public_profile", cardinality: "many-to-many"}

```

This is `creatorsError.hint`:
```
Try changing 'public_profile' to one of the following: 'public_profile!adaptation_notifications', 'public_profile!adaptation_notifications', 'public_profile!creator_follows', 'public_profile!creator_follows'. Find the desired relationship in the 'details' key.
```

Can I get some help?

**Comment by steve-chavez** [comment]
@Electric-Dragon Could you share the full error? There should be more details in the "details" field.

**Reply by Electric-Dragon** [reply]
@steve-chavez Sorry but that's all that comes up in the `details` field

**Reply by steve-chavez** [reply]
@Electric-Dragon Sorry, I've just noted that you already shared the `details` field :sweat_smile:

**Reply by Electric-Dragon** [reply]
Yes😅

**Reply by steve-chavez** [reply]
@Electric-Dragon Ok, I've just noted that https://github.com/PostgREST/postgrest/pull/2262 will fix your issue. That change is not in production yet, but if you email me(to steve@supabase.io) your project ref I can patch your instance with the change.

**Reply by Electric-Dragon** [reply]
@steve-chavez Okay thanks a lot. I've done the needful. Also, do you know what it means to get this error: `Cannot apply filter because 'series' is not an embedded resource in this request`?

**Reply by steve-chavez** [reply]
@Electric-Dragon Can you share the request for that one?

**Reply by Electric-Dragon** [reply]
@steve-chavez I tried changing some things and this is the request:

```
        const {data:popularSeries, error:popularSeriesError} = await supabase
                .from('series_popularity')
                .select('series!inner(cover,title,id,creator:public_profile!series_creator_fkey(username))')
                .order('popularity_score', { ascending: false })
                .or(`genre1.eq.${genre},genre2.eq.${genre}`, { foreignTable: "series" })
                .neq('series.status', 'd')
                .limit(6)
```

This is the new error: `{message: "column \"series\" does not exist", code: "42703", details: null, hint: null}`

---

View definition for series_popularity:

```
create or replace view series_popularity as (

  select series_total_reads.series_id, ((0.5 * get_series_follows(series_total_reads.series_id)) + (0.1 * total_read_count) + (0.4 * series_total_likes.count)) as popularity_score

  from series_total_reads

  join series_total_likes on series_total_likes.seriesid = series_total_reads.series_id

)
```

**Reply by steve-chavez** [reply]
@Electric-Dragon Sorry, I'm a bit confused. If that's a different issue than the one in the [above discussion](https://github.com/supabase/supabase/discussions/6848#discussion-4076525) could you open another discussion?

**Reply by Electric-Dragon** [reply]
Okay I shall do that

**Reply by Electric-Dragon** [reply]
@steve-chavez I have done the needful in #6860

**Comment by steve-chavez** [comment, marked as answer]
@Electric-Dragon I've patched your instance with the [latest postgREST pre-release](https://github.com/PostgREST/postgrest/releases/tag/v9.0.0.20220516), let me know if that fixes this issue.

**Reply by Electric-Dragon** [reply]
@steve-chavez This is the error I get now

<img width="1440" alt="Bildschirmfoto 2022-05-17 um 10 38 47 AM" src="https://user-images.githubusercontent.com/45729391/168733567-c35cea0f-519a-4efc-8364-5ea3d593a3c1.png">

**Reply by steve-chavez** [reply]
@Electric-Dragon Ok. So, looking at your view definition. Does `public_profile` has a relationship to itself? Or is there a relation from `creator_total_likes` to `public_profile`?

**Reply by steve-chavez** [reply]
Just checked and I think it should be the latter. @Electric-Dragon Can you post the definition of the `creator_total_likes` table?

**Reply by Electric-Dragon** [reply]
Here you go:

```
create or replace view creator_total_likes as (
      select public_profile.id as creator_id, sum(series_total_likes.count) as total_likes

      from public_profile

      join series on public_profile.id = series.creator

      join series_total_likes

      on series_total_likes.seriesid = series.id

      group by public_profile.id
)
```

**Reply by steve-chavez** [reply]
@Electric-Dragon Hm. So embedding views works by inferring the base tables of the views and then checking if these base tables have relationships between them.([ref](https://postgrest.org/en/stable/api.html#embedding-views))

`creator_ranks` has `public_profile` as its base table(the one in the `FROM public_profile`) and you're trying to embed that with `public_profile`, but there's no relationship from `public_profile` to `public_profile`.

Assuming `creator_ranks` had `creator_total_likes` as its base, then `creator_total_likes` has `public_profile` as its base table, which would give the same problem as above - no relationship from `public_profile` to `public_profile`.

How would you expect to join `creator_ranks` with `public_profile` through plain SQL?

**Reply by Electric-Dragon** [reply]
@steve-chavez  I don't get what you mean by:
> there's no relationship from public_profile to public_profile

**Reply by steve-chavez** [reply]
@Electric-Dragon I mean a foreign key relationship, there isn't one defined from `public_profile` to the same `public_profile`.

**Reply by Electric-Dragon** [reply]
@steve-chavez  So should I reference `public_profile.id` to `public_profile.id`?

**Reply by steve-chavez** [reply]
@Electric-Dragon Oh, no. Not suggesting that. Is just that we need foreign keys to have relationships.

Looking further into your view, perhaps you're expecting to join based on `series.creator`?

Does `series` have a foreign key relationship to `public_profile`?

**Reply by Electric-Dragon** [reply]
@steve-chavez  It does.

`series.creator` refences `public_profile.id`

**Reply by steve-chavez** [reply]
@Electric-Dragon Ok, then I think we should do this for your original snippet to work. We need to change your views definition to:

```sql
create or replace view creator_total_likes as (
      select series.creator as creator_id, sum(series_total_likes.count) as total_likes

      from public_profile

      join series on public_profile.id = series.creator

      join series_total_likes

      on series_total_likes.seriesid = series.id

      group by public_profile.id
)
-- Should be the same resultset, basically changing `public_profile.id` to `series.creator`

create or replace view creator_ranks as (
	SELECT
		creator_total_likes.creator_id as creator_id,
		RANK() OVER (
			ORDER BY ((0.3 * total_likes) + (0.5 * get_creator_follows(creator_id)) + (0.2 * get_creator_total_reads(creator_id)))
		) creator_rank

	FROM public_profile

	JOIN creator_total_likes on public_profile.id = creator_total_likes.creator_id

)
-- Same resultset as well, changing `public_profile.id` to `creator_total_likes.creator_id `
```

Then retry:

```js
         const { data:creators, error:creatorsError } = await supabase
             .from('creator_ranks')
             .select('public_profile!inner(username,pfp,creator_type)')
             .order('creator_rank', { ascending: false })
             .limit(5)
```

**Reply by Electric-Dragon** [reply]
This work...thanks a lot!! So if I'm correct, the base table should not be `public_profile` if I'm trying to access data from it using an inner join?

**Reply by steve-chavez** [reply]
Some postgREST docs are coming to explain this. 

> So if I'm correct, the base table should not be public_profile if I'm trying to access data from it using an inner join?

First of all, using an inner or not doesn't matter for this case - doing `select('public_profile(username,pfp,creator_type)')` should also work.

To put it generally, postgREST will infer the base table from the table's columns in the SELECT of the view definition.

What I did here is to adjust the views to let postgREST find the relationship you mentioned [above](https://github.com/supabase/supabase/discussions/6848#discussioncomment-2769615):

> series.creator refences public_profile.id

For this, the `creator_total_likes` inferred base table should be `series`, so I adjusted that by changing ` public_profile.id` to `series.creator` in the VIEW definition.

Then changing `creator_ranks` inferred base view to `creator_total_likes`, by changing `public_profile.id` to `creator_total_likes.creator_id `, this way `creator_ranks` has `series` as its base table(two levels down) and finally it has a relationship to `public_profile`.

**Reply by Electric-Dragon** [reply]
Ahh Okay..Thanks a lott!
